### PR TITLE
CorrelationIdGenerator uses string.Create and avoids bounds-checks

### PR DESCRIPTION
**Pull request by gfoidl**
# Description

For `netcoreapp2.1` `string.Create` is used instead of a stack allocation and then copying to the string.

The bound-checks are removed for both targetframeworks.

# Benchmarks

``` ini

BenchmarkDotNet=v0.10.14, OS=Windows 10.0.17134
Intel Core i7-7700HQ CPU 2.80GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
Frequency=2742192 Hz, Resolution=364.6718 ns, Timer=TSC
.NET Core SDK=2.1.300-rc1-008673
  [Host] : .NET Core 2.1.0-rc1 (CoreCLR 4.6.26426.02, CoreFX 4.6.26426.04), 64bit RyuJIT
  Clr    : .NET Framework 4.7.1 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.7.3101.0
  Core   : .NET Core 2.1.0-rc1 (CoreCLR 4.6.26426.02, CoreFX 4.6.26426.04), 64bit RyuJIT


```
|                 Method |  Job | Runtime |     Mean |     Error |    StdDev | Scaled |  Gen 0 | Allocated |
|----------------------- |----- |-------- |---------:|----------:|----------:|-------:|-------:|----------:|
|                Default |  Clr |     Clr | 30.55 ns | 0.3343 ns | 0.2610 ns |   1.00 | 0.0178 |      56 B |
|           StringCreate |  Clr |     Clr | 30.72 ns | 0.3675 ns | 0.3437 ns |   1.01 | 0.0178 |      56 B |
| StringCreateSpanUnsafe |  Clr |     Clr | 29.00 ns | 0.1565 ns | 0.1463 ns |   0.95 | 0.0178 |      56 B |
|                        |      |         |          |           |           |        |        |           |
|                Default | Core |    Core | 38.56 ns | 0.7330 ns | 0.6498 ns |   1.00 | 0.0178 |      56 B |
|           StringCreate | Core |    Core | 29.07 ns | 0.1224 ns | 0.1022 ns |   0.75 | 0.0178 |      56 B |
| StringCreateSpanUnsafe | Core |    Core | 24.21 ns | 0.1273 ns | 0.1129 ns |   0.63 | 0.0178 |      56 B |

[Default](https://github.com/gfoidl/Benchmarks/blob/c14386b94f74389bc57d0b36d2c2d63f0a1b0b26/aspnet/Kestrel/CorrelationIdGenerator/CorrelationIdGenerator/CorrelationIdGenerators/CorrelationIdGenerator.cs) is the current implementation in _dev_.
[StringCreate](https://github.com/gfoidl/Benchmarks/blob/c14386b94f74389bc57d0b36d2c2d63f0a1b0b26/aspnet/Kestrel/CorrelationIdGenerator/CorrelationIdGenerator/CorrelationIdGenerators/CorrelationIdGenerator1.cs) is for reference when just `string.Create` is used (only applyable for `netcoreapp2.1`.
[StringCreateSpanUnsafe](https://github.com/gfoidl/Benchmarks/blob/c14386b94f74389bc57d0b36d2c2d63f0a1b0b26/aspnet/Kestrel/CorrelationIdGenerator/CorrelationIdGenerator/CorrelationIdGenerators/CorrelationIdGenerator2.cs) is this PR.

### CODE REVIEW

**Comment by gfoidl** [comment]
I just saw some other places where the bounds-check could be avoided in the same way, like

https://github.com/aspnet/KestrelHttpServer/blob/8da5f55832acf0386682726925112ebd411db60c/src/Kestrel.Core/Internal/Http/ChunkWriter.cs#L24-L33

Shall I do this? (if so, in this PR or a new one?)

File: src/Kestrel.Core/Internal/Infrastructure/CorrelationIdGenerator.cs [hunk 1]
```diff
@@ -24,25 +30,48 @@ private static unsafe string GenerateId(long id)
...
+        {
+            ref char buffer = ref MemoryMarshal.GetReference(charBuffer);
+            ref char encodeChars = ref _encode32Chars[0];
+
+            FillBuffer(ref buffer, id, ref encodeChars);
+        }
+#endif
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void FillBuffer(ref char buffer, long id, ref char encodeChars)
```
Comment (davidfowl) [inline src/Kestrel.Core/Internal/Infrastructure/CorrelationIdGenerator.cs]: Can't you get rid of the bounds check without using unsafe?

**Comment by halter73** [comment]
Thanks! This is a nice improvement on Core without regressing Desktop or complicating the code too much.

In addition to eliminating bounds checks ChunkWriter.BeginChunkBytes, you might also want to look at StringUtilities.ConcatAsHexSuffix.

Before you spend too much more time, I'd like some feedback from @benaadams who's responsible for the majority of the unsafe code in Kestrel. LGTM though.

File: src/Kestrel.Core/Internal/Infrastructure/CorrelationIdGenerator.cs (hunk 1 above)
Comment (gfoidl) [reply src/Kestrel.Core/Internal/Infrastructure/CorrelationIdGenerator.cs]: Hm, I don't know how.

Even if the `char[]`-field (`_encode32Chars`) will be assigned to a local, the JIT has to emit a bound-check due the "random" access to this array.

For the `charBuffer` it isn't a problem.

File: src/Kestrel.Core/Internal/Infrastructure/CorrelationIdGenerator.cs [hunk 2]
```diff
@@ -24,25 +30,48 @@ private static unsafe string GenerateId(long id)
...
-            charBuffer[12] = _encode32Chars[(int)id & 31];
+            FillBuffer(ref Unsafe.AsRef<char>(charBuffer), id, ref encodeChars);
 
             // string ctor overload that takes char*
             return new string(charBuffer, 0, 13);
+#endif
+        }
+
+#if NETCOREAPP2_1
+        private static readonly SpanAction<char, long> _spanAction = FillBufferAction;
```
Comment (halter73) [inline src/Kestrel.Core/Internal/Infrastructure/CorrelationIdGenerator.cs]: Nit: move to top of class.

File: src/Kestrel.Core/Internal/Infrastructure/CorrelationIdGenerator.cs [hunk 3]
```diff
@@ -3,13 +3,19 @@
 
 using System;
 using System.Threading;
+using System.Runtime.CompilerServices;
+
+#if NETCOREAPP2_1
+using System.Buffers;
+using System.Runtime.InteropServices;
+#endif
```
Comment (halter73) [inline src/Kestrel.Core/Internal/Infrastructure/CorrelationIdGenerator.cs]: Does this need to be ifdefed?

Comment (gfoidl) [reply src/Kestrel.Core/Internal/Infrastructure/CorrelationIdGenerator.cs]: No.

It's an artefact from my personal code-style. Shall I remove this?

**Comment by gfoidl** [comment]
> you might also want to look at 

I'll have a look over the place. Thx.

File: src/Kestrel.Core/Internal/Infrastructure/CorrelationIdGenerator.cs [hunk 4]
```diff
@@ -24,25 +30,48 @@ private static unsafe string GenerateId(long id)
...
-            charBuffer[4] = _encode32Chars[(int)(id >> 40) & 31];
-            charBuffer[5] = _encode32Chars[(int)(id >> 35) & 31];
-            charBuffer[6] = _encode32Chars[(int)(id >> 30) & 31];
-            charBuffer[7] = _encode32Chars[(int)(id >> 25) & 31];
-            charBuffer[8] = _encode32Chars[(int)(id >> 20) & 31];
-            charBuffer[9] = _encode32Chars[(int)(id >> 15) & 31];
-            charBuffer[10] = _encode32Chars[(int)(id >> 10) & 31];
-            charBuffer[11] = _encode32Chars[(int)(id >> 5) & 31];
-            charBuffer[12] = _encode32Chars[(int)id & 31];
+            FillBuffer(ref Unsafe.AsRef<char>(charBuffer), id, ref encodeChars);
```
Comment (halter73) [inline src/Kestrel.Core/Internal/Infrastructure/CorrelationIdGenerator.cs]: Would `FillBuffer(ref Unsafe.AsRef<char>(charBuffer), id, ref _encode32Chars[0])` produce less efficient code? It's fine the way it is, but I'm now paranoid that the local is somehow necessary for JIT to optimize away the bounds check.

File: src/Kestrel.Core/Internal/Infrastructure/CorrelationIdGenerator.cs (hunk 3 above)
Comment (halter73) [reply src/Kestrel.Core/Internal/Infrastructure/CorrelationIdGenerator.cs]: Not a big deal, but yes.

File: src/Kestrel.Core/Internal/Infrastructure/CorrelationIdGenerator.cs (hunk 4 above)
Comment (gfoidl) [reply src/Kestrel.Core/Internal/Infrastructure/CorrelationIdGenerator.cs]: > that the local is somehow necessary for JIT to optimize away the bounds check

The bounds-check for `_encode32Chars[0]` won't be eliminated either way :sad:
Though I assume the JIT could to this on `static readonly`s. Don't know if this tracked anywhere.

I'll update it to the "inline" version w/o the local.

**Comment by gfoidl** [comment]
### StringUtilities.ConcatAsHexSuffix

``` ini

BenchmarkDotNet=v0.10.14, OS=Windows 10.0.17134
Intel Core i7-7700HQ CPU 2.80GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
Frequency=2742192 Hz, Resolution=364.6718 ns, Timer=TSC
.NET Core SDK=2.1.300-rc1-008673
  [Host] : .NET Core 2.1.0-rc1 (CoreCLR 4.6.26426.02, CoreFX 4.6.26426.04), 64bit RyuJIT
  Clr    : .NET Framework 4.7.1 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.7.3101.0
  Core   : .NET Core 2.1.0-rc1 (CoreCLR 4.6.26426.02, CoreFX 4.6.26426.04), 64bit RyuJIT


```
|                 Method |  Job | Runtime |     Mean |     Error |    StdDev | Scaled | ScaledSD |  Gen 0 | Allocated |
|----------------------- |----- |-------- |---------:|----------:|----------:|-------:|---------:|-------:|----------:|
|                Default |  Clr |     Clr | 31.33 ns | 0.6917 ns | 0.8748 ns |   1.00 |     0.00 | 0.0228 |      72 B |
| StringCreateSpanUnsafe |  Clr |     Clr | 30.18 ns | 0.2140 ns | 0.1897 ns |   0.96 |     0.03 | 0.0228 |      72 B |
|                        |      |         |          |           |           |        |          |        |           |
|                Default | Core |    Core | 29.77 ns | 0.2383 ns | 0.2112 ns |   1.00 |     0.00 | 0.0228 |      72 B |
| StringCreateSpanUnsafe | Core |    Core | 24.77 ns | 0.1765 ns | 0.1651 ns |   0.83 |     0.01 | 0.0229 |      72 B |

**Comment by gfoidl** [comment]
Other places I've found, but don't know how relevant they are (and didn't try anything so far, just a search in VS):

https://github.com/aspnet/KestrelHttpServer/blob/27799b5acb59523889a3a52a39d606e795b81c50/src/Kestrel.Core/Internal/Http2/HPack/Huffman.cs#L380

https://github.com/aspnet/KestrelHttpServer/blob/27799b5acb59523889a3a52a39d606e795b81c50/src/Kestrel.Core/Internal/Http2/Http2Frame.cs#L20

**Comment by halter73** [comment]
HTTP/2 isn't supported yet and I expect it will undergo a lot of churn in 2.2. I'd leave it as is for now. I'll keep in mind that it's a possible avenue for improving HTTP/2 perf.

File: src/Kestrel.Core/Internal/Infrastructure/StringUtilities.cs [hunk 5]
```diff
@@ -126,31 +131,46 @@ public static unsafe string ConcatAsHexSuffix(string str, char separator, uint n
...
+        {
             var i = 0;
+
             if (str != null)
             {
-                for (i = 0; i < str.Length; i++)
+                for (; i < str.Length; ++i)
                 {
-                    charBuffer[i] = str[i];
+                    Unsafe.Add(ref charBuffer, i) = str[i];
```
[the remark this answers is not in the document]
Comment (benaadams) [reply src/Kestrel.Core/Internal/Infrastructure/StringUtilities.cs]: Have raised issue for ROSpan.CopyTo https://github.com/dotnet/coreclr/issues/18005

File: src/Kestrel.Core/Internal/Http/ChunkWriter.cs [hunk 6]
```diff
@@ -21,19 +22,20 @@ private static ArraySegment<byte> CreateAsciiByteArraySegment(string text)
...
-                _hex[((dataCount >> 0x04) & 0x0f)],
-                _hex[((dataCount >> 0x00) & 0x0f)],
-                (byte)'\r',
-                (byte)'\n',
-            };
+            var bytes = new byte[10];
+            ref var r = ref bytes[0];
+            ref var hex = ref _hex[0];
+
+            Unsafe.Add(ref r, 0) = Unsafe.Add(ref hex, (dataCount >> 0x1c) & 0x0f);
```
Comment (jkotas) [inline src/Kestrel.Core/Internal/Http/ChunkWriter.cs]: You should not need any of this unsafe code. This should be able to eliminate the bounds checks just fine.

Comment (jkotas) [reply src/Kestrel.Core/Internal/Http/ChunkWriter.cs]: (for bytes). For `_hex` - not yet.

Comment (benaadams) [reply src/Kestrel.Core/Internal/Http/ChunkWriter.cs]: Could drop the `_hex` loopkup table and use?

```csharp
// digit = (dataCount >> 0x1c) & 0x0f;
= (byte)(digit + (digit < 10 ? (byte)'0' : (byte)'a'));
```

Comment (gfoidl) [reply src/Kestrel.Core/Internal/Http/ChunkWriter.cs]: @jkotas thx! I got distracted by the generated `jae` and `jae SHORT`.

For `_hex` I'll try the idea from @benaadams

Comment (gfoidl) [reply src/Kestrel.Core/Internal/Http/ChunkWriter.cs]:
``` ini

BenchmarkDotNet=v0.10.14, OS=Windows 10.0.17134
Intel Core i7-7700HQ CPU 2.80GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
Frequency=2742192 Hz, Resolution=364.6718 ns, Timer=TSC
.NET Core SDK=2.1.300-rc1-008673
  [Host]     : .NET Core 2.1.0-rc1 (CoreCLR 4.6.26426.02, CoreFX 4.6.26426.04), 64bit RyuJIT
  DefaultJob : .NET Core 2.1.0-rc1 (CoreCLR 4.6.26426.02, CoreFX 4.6.26426.04), 64bit RyuJIT


```
|             Method |     Mean |     Error |    StdDev | Scaled | ScaledSD |  Gen 0 | Allocated |
|------------------- |---------:|----------:|----------:|-------:|---------:|-------:|----------:|
|            Default | 17.51 ns | 0.3823 ns | 0.3755 ns |   1.00 |     0.00 | 0.0127 |      40 B |
|             Unsafe | 15.43 ns | 0.1269 ns | 0.1187 ns |   0.88 |     0.02 | 0.0127 |      40 B |
| BytesSafeHexUnsafe | 13.88 ns | 0.1718 ns | 0.1607 ns |   0.79 |     0.02 | 0.0127 |      40 B |
|           NoLookup | 14.28 ns | 0.1124 ns | 0.0938 ns |   0.82 |     0.02 | 0.0127 |      40 B |
| NoLookupBranchless | 16.28 ns | 0.3119 ns | 0.2918 ns |   0.93 |     0.03 | 0.0127 |      40 B |

`Default` is _dev_-branch
`Unsafe` this PR so far
[BytesSafeHexUnsafe](https://github.com/gfoidl/Benchmarks/blob/0fe785ac88273e977adf44c1dcd8891651316420/aspnet/Kestrel/CorrelationIdGenerator_PR2579/CorrelationIdGenerator/ChunkWriters/ChunkWriter2.cs) doesn't use `Unsafe` for `bytes`
[NoLookup](https://github.com/gfoidl/Benchmarks/blob/0fe785ac88273e977adf44c1dcd8891651316420/aspnet/Kestrel/CorrelationIdGenerator_PR2579/CorrelationIdGenerator/ChunkWriters/ChunkWriter3.cs) is @benaadams idea
[NoLookupBranchless](https://github.com/gfoidl/Benchmarks/blob/0fe785ac88273e977adf44c1dcd8891651316420/aspnet/Kestrel/CorrelationIdGenerator_PR2579/CorrelationIdGenerator/ChunkWriters/ChunkWriter4.cs) is @benaadams idea in a branchless variant.

I've update the PR to use `NoLookup` even if it minimal slower than `BytesSafeHexUnsafe`, but there's no unsafe anymore.

For me it's interesting that the branchless version is slower than a branching version (seems that the branch-predictor does a good job):
```asm
; version with branches
; ...
       8BFB                 mov      edi, ebx
       C1FF18               sar      edi, 24
       83E70F               and      edi, 15
       83FF0A               cmp      edi, 10
       7C07                 jl       SHORT G_M56191_IG05
       BE57000000           mov      esi, 87
       EB05                 jmp      SHORT G_M56191_IG06

G_M56191_IG05:
       BE30000000           mov      esi, 48

G_M56191_IG06:
       03FE                 add      edi, esi
       400FB6FF             movzx    rdi, dil
       40887811             mov      byte  ptr [rax+17], dil
; ...

; branchless version
; ...
       8BFB                 mov      edi, ebx
       C1FF18               sar      edi, 24
       83E70F               and      edi, 15
       8D77F6               lea      esi, [rdi-10]
       C1FE1F               sar      esi, 31
       83E667               and      esi, 103
       83F657               xor      esi, 87
       03FE                 add      edi, esi
       400FB6FF             movzx    rdi, dil
       40887811             mov      byte  ptr [rax+17], dil
; ...
```

Though I would prefer anyway the version with the ternary and let the JIT emit proper code.

**Comment by gfoidl** [comment]
> I'll keep in mind that it's a possible avenue for improving HTTP/2 perf.

@halter73 fine, thx. In case of let me know if I can help.

File: src/Kestrel.Core/Internal/Http/ChunkWriter.cs (hunk 6 above)
Comment (gfoidl) [reply src/Kestrel.Core/Internal/Http/ChunkWriter.cs]: I'll try to do the same for `CorrelationIdGenerator` and `StringUtilities`.

**Comment by gfoidl** [comment]
### CorrelationIdGenerator.GetNextId w/o lookup (PR not updated yet)

``` ini

BenchmarkDotNet=v0.10.14, OS=Windows 10.0.17134
Intel Core i7-7700HQ CPU 2.80GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
Frequency=2742192 Hz, Resolution=364.6718 ns, Timer=TSC
.NET Core SDK=2.1.300-rc1-008673
  [Host] : .NET Core 2.1.0-rc1 (CoreCLR 4.6.26426.02, CoreFX 4.6.26426.04), 64bit RyuJIT
  Clr    : .NET Framework 4.7.1 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.7.3101.0
  Core   : .NET Core 2.1.0-rc1 (CoreCLR 4.6.26426.02, CoreFX 4.6.26426.04), 64bit RyuJIT


```
|                 Method |  Job | Runtime |     Mean |     Error |    StdDev | Scaled | ScaledSD |  Gen 0 | Allocated |
|----------------------- |----- |-------- |---------:|----------:|----------:|-------:|---------:|-------:|----------:|
|                Default |  Clr |     Clr | 32.91 ns | 0.6418 ns | 0.6004 ns |   1.00 |     0.00 | 0.0178 |      56 B |
|           StringCreate |  Clr |     Clr | 31.06 ns | 0.4951 ns | 0.4631 ns |   0.94 |     0.02 | 0.0178 |      56 B |
| StringCreateSpanUnsafe |  Clr |     Clr | 28.88 ns | 0.3857 ns | 0.3608 ns |   0.88 |     0.02 | 0.0178 |      56 B |
|               NoLookup |  Clr |     Clr | 32.77 ns | 0.4607 ns | 0.4309 ns |   1.00 |     0.02 | 0.0178 |      56 B |
|     NoLookupBranchless |  Clr |     Clr | 35.48 ns | 0.1829 ns | 0.1428 ns |   1.08 |     0.02 | 0.0178 |      56 B |
|                        |      |         |          |           |           |        |          |        |           |
|                Default | Core |    Core | 38.00 ns | 0.2891 ns | 0.2705 ns |   1.00 |     0.00 | 0.0178 |      56 B |
|           StringCreate | Core |    Core | 29.55 ns | 0.3055 ns | 0.2858 ns |   0.78 |     0.01 | 0.0178 |      56 B |
| StringCreateSpanUnsafe | Core |    Core | 24.50 ns | 0.2452 ns | 0.2294 ns |   0.64 |     0.01 | 0.0178 |      56 B |
|               NoLookup | Core |    Core | 27.75 ns | 0.4264 ns | 0.3780 ns |   0.73 |     0.01 | 0.0178 |      56 B |
|     NoLookupBranchless | Core |    Core | 28.91 ns | 0.4621 ns | 0.4322 ns |   0.76 |     0.01 | 0.0178 |      56 B |

[               Default](https://github.com/gfoidl/Benchmarks/blob/3ec34cebbfc5b09299fe227e35d0e86e307bc615/aspnet/Kestrel/CorrelationIdGenerator_PR2579/CorrelationIdGenerator/CorrelationIdGenerators/CorrelationIdGenerator.cs) is the current implementation in _dev_.
[          StringCreate](https://github.com/gfoidl/Benchmarks/blob/3ec34cebbfc5b09299fe227e35d0e86e307bc615/aspnet/Kestrel/CorrelationIdGenerator_PR2579/CorrelationIdGenerator/CorrelationIdGenerators/CorrelationIdGenerator1.cs) is for reference when just `string.Create` is used (only applyable for `netcoreapp2.1`.
[StringCreateSpanUnsafe](https://github.com/gfoidl/Benchmarks/blob/3ec34cebbfc5b09299fe227e35d0e86e307bc615/aspnet/Kestrel/CorrelationIdGenerator_PR2579/CorrelationIdGenerator/CorrelationIdGenerators/CorrelationIdGenerator2.cs) is this PR so far.
[              NoLookup](https://github.com/gfoidl/Benchmarks/blob/3ec34cebbfc5b09299fe227e35d0e86e307bc615/aspnet/Kestrel/CorrelationIdGenerator_PR2579/CorrelationIdGenerator/CorrelationIdGenerators/CorrelationIdGenerator3.cs) w/o lookup, `Unsafe` only on `buffer`
[    NoLookupBranchless](https://github.com/gfoidl/Benchmarks/blob/3ec34cebbfc5b09299fe227e35d0e86e307bc615/aspnet/Kestrel/CorrelationIdGenerator_PR2579/CorrelationIdGenerator/CorrelationIdGenerators/CorrelationIdGenerator4.cs) w/o lookup and branchless (for reference)

So `StringCreateSpanUnsafe` is the current state of this PR.
`NoLookup` would be the update. Perf regresses a bit (both Core and Desktop). I would stick to `StringCreateSpanUnsafe`, but shall I update to `NoLookup`?

**Comment by gfoidl** [comment]
### StringUtilities.ConcatAsHexSuffix w/o lookup (PR not updated yet)

``` ini

BenchmarkDotNet=v0.10.14, OS=Windows 10.0.17134
Intel Core i7-7700HQ CPU 2.80GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
Frequency=2742192 Hz, Resolution=364.6718 ns, Timer=TSC
.NET Core SDK=2.1.300-rc1-008673
  [Host] : .NET Core 2.1.0-rc1 (CoreCLR 4.6.26426.02, CoreFX 4.6.26426.04), 64bit RyuJIT
  Clr    : .NET Framework 4.7.1 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.7.3101.0
  Core   : .NET Core 2.1.0-rc1 (CoreCLR 4.6.26426.02, CoreFX 4.6.26426.04), 64bit RyuJIT


```
|                 Method |  Job | Runtime |     Mean |     Error |    StdDev | Scaled | ScaledSD |  Gen 0 | Allocated |
|----------------------- |----- |-------- |---------:|----------:|----------:|-------:|---------:|-------:|----------:|
|                Default |  Clr |     Clr | 31.14 ns | 0.3393 ns | 0.3173 ns |   1.00 |     0.00 | 0.0228 |      72 B |
| StringCreateSpanUnsafe |  Clr |     Clr | 30.45 ns | 0.5382 ns | 0.5034 ns |   0.98 |     0.02 | 0.0228 |      72 B |
|               NoLookup |  Clr |     Clr | 33.77 ns | 0.4602 ns | 0.4304 ns |   1.08 |     0.02 | 0.0228 |      72 B |
|     NoLookupBranchless |  Clr |     Clr | 33.81 ns | 0.4356 ns | 0.3861 ns |   1.09 |     0.02 | 0.0228 |      72 B |
|                        |      |         |          |           |           |        |          |        |           |
|                Default | Core |    Core | 34.12 ns | 0.7521 ns | 0.7386 ns |   1.00 |     0.00 | 0.0228 |      72 B |
| StringCreateSpanUnsafe | Core |    Core | 26.85 ns | 0.4039 ns | 0.3778 ns |   0.79 |     0.02 | 0.0228 |      72 B |
|               NoLookup | Core |    Core | 26.77 ns | 0.3949 ns | 0.3694 ns |   0.79 |     0.02 | 0.0228 |      72 B |
|     NoLookupBranchless | Core |    Core | 28.02 ns | 0.3557 ns | 0.3328 ns |   0.82 |     0.02 | 0.0228 |      72 B |

[               Default](https://github.com/gfoidl/Benchmarks/blob/bfae3bcf72d9e6804f91114027ae5d9562bdc9f1/aspnet/Kestrel/CorrelationIdGenerator_PR2579/CorrelationIdGenerator/StringUtilitiess/StringUtilities.cs#L121) is the current implementation in _dev_.
[StringCreateSpanUnsafe](https://github.com/gfoidl/Benchmarks/blob/bfae3bcf72d9e6804f91114027ae5d9562bdc9f1/aspnet/Kestrel/CorrelationIdGenerator_PR2579/CorrelationIdGenerator/StringUtilitiess/StringUtilities1.cs#L126) is this PR so far.
[              NoLookup](https://github.com/gfoidl/Benchmarks/blob/bfae3bcf72d9e6804f91114027ae5d9562bdc9f1/aspnet/Kestrel/CorrelationIdGenerator_PR2579/CorrelationIdGenerator/StringUtilitiess/StringUtilities2.cs#L125) w/o lookup, `Unsafe` only on `charBuffer`
[    NoLookupBranchless](https://github.com/gfoidl/Benchmarks/blob/bfae3bcf72d9e6804f91114027ae5d9562bdc9f1/aspnet/Kestrel/CorrelationIdGenerator_PR2579/CorrelationIdGenerator/StringUtilitiess/StringUtilities3.cs#L125) w/o lookup and branchless (for reference)

So `StringCreateSpanUnsafe` is the current state of this PR.
`NoLookup` would be the update. On Core perf is equal, on Desktop perf regresses. I would update the PR to use `NoLookup`. Shall I?

File: src/Kestrel.Core/Internal/Infrastructure/CorrelationIdGenerator.cs (hunk 1 above)
Comment (davidfowl) [reply src/Kestrel.Core/Internal/Infrastructure/CorrelationIdGenerator.cs]: TBH, using unsafe code for this little of a gain isn't very attractive. We want to start removing unsafe code from Kestrel and using span in more places.

Comment (gfoidl) [reply src/Kestrel.Core/Internal/Infrastructure/CorrelationIdGenerator.cs]: > using span in more places.

Is definitely good. 👍 

But as long the JIT has to emit bound-checks for span there will be some unsafe parts. Span and unsafe is not xor :wink: I know what you mean...thx.

**Comment by muratg** [comment]
We’re closing this PR because we don’t feel that this change is a good fit for the product at this time - we don't want to add more unsafe code to Kestrel.

We thank you for the contribution and look forward to collaborating more in the future.

**Pull request closed without merging**